### ISSUE TITLE
Documentation: Understanding ExponentiatedGradient' termination and guarantees

### ISSUE THREAD

**Issue by kevinrobinson** [fairlearn/fairlearn#444]
I'm trying to better understand what leads `ExponentiatedGradient` to terminate.  In particular, I'm trying to understand this in terms that I could translate to a non-technical person who wants to understand if they can control how much slack there is around the selection rates being exactly equal.  To make this concrete, I'm looking at the example in the quickstart guide, where a `DemographicParity` constraint for the `sex` column is applied to a task predicting which people earned over $50k.  I'd like to understand how to apply the `ExponentiatedGradient` so that it will enforce that the value of `demographic_parity_difference` is < 0.05 across the male and female groups.

I looked into understanding how the `eps` and `nu` params influence this.  I also noticed the `_MIN_ITER` check, and that the example scenario would terminate earlier if this is disabled.

To experiment, I set `eps=0.05` but found that this produced a demographic parity difference of `0.072`, so outside of the bounds I was expecting.  Similarly, setting `eps=0.10` produced a demographic disparity of `0.149` which was out of the bounds of what I was expecting.  So I'm figuring I'm misunderstanding something important here.

If it's helpful, this is the code from the quickstart I'm starting from, and an example of what that case shows with debug logging enabled.  The examples above all use the same `seed(0)` from the quickstart.

```
np.random.seed(0)
constraint = DemographicParity()
classifier = DecisionTreeClassifier()
mitigator = ExponentiatedGradient(classifier, constraint, eps=whatever)
mitigator.fit(X, y_true, sensitive_features=sex)
y_pred_mitigated = mitigator.predict(X)
...
Classification problem detected
...Exponentiated Gradient STARTING
...iter=000
...eps=0.010, B=100.0, nu=0.000421, max_iter=50, eta_min=0.000002
         eta=0.020000, L_low=0.024, L=0.024, L_high=10.862, gap=10.837989, disp=0.118, err=0.024, gap_LP=inf
...iter=001
         eta=0.020000, L_low=0.030, L=0.030, L_high=10.652, gap=10.622350, disp=0.116, err=0.024, gap_LP=77.863978
...iter=002
         eta=0.020000, L_low=0.036, L=0.036, L_high=10.486, gap=10.450200, disp=0.115, err=0.024, gap_LP=0.069860
...iter=003
         eta=0.020000, L_low=0.041, L=0.042, L_high=10.248, gap=10.206413, disp=0.112, err=0.024, gap_LP=0.000108
...iter=004
         eta=0.020000, L_low=0.047, L=0.047, L_high=10.091, gap=10.043995, disp=0.111, err=0.025, gap_LP=0.000000
...iter=005
         eta=0.020000, L_low=0.052, L=0.074, L_high=0.133, gap=0.059818, disp=0.011, err=0.074, gap_LP=0.000000
...eps=0.010, B=100.0, nu=0.000421, max_iter=50, eta_min=0.000002
...last_iter=5, best_iter=5, best_gap=0.000000, n_oracle_calls=26, n_hs=12
```

And I am running this to see what the demographic parity difference is afterward:
```
print(metrics.demographic_parity_difference(y_true, y_pred_mitigated, sensitive_features=sex))
0.010575424150881252
```

Thanks! :)

**Comment by kevinrobinson** [comment]
Ah I also found [this description](https://github.com/fairlearn/fairlearn/blob/master/fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py#L34) afterward:

>  eps: Allowed fairness constraint violation; the solution is guaranteed to have the error within 2*best_gap of the best error under constraint eps; the constraint violation is at most 2*(eps+best_gap)

This seems important :)  

In reading the code, this doesn't seem to match what I see in the [line](https://github.com/fairlearn/fairlearn/blob/master/fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py#L153):

>   `  if (gaps[t] < self.nu) and (t >= _MIN_ITER)` 

This matches my understanding of what I've seen in the paper - that there is no "best gap" in this method since we're stopping early as soon as the solution is tolerable in terms of nu.

So I'm still trying to understand these params and how to control them at the abstraction level of "enforce demographic parity so that when we run `metrics.demographic_parity_difference` afterward, it will be below 0.05."  (I'm also aware that these questions are only focused on understanding how the constraints are enforced on the training dataset, and we're not even talking about generalization yet.)

**Comment by kevinrobinson** [comment]
From reading the Moments code I thought I had figured out that from an API perspective that `DemographicParity(0.05)` was the way to ask for something like this, but I was surprised to discover that this didn't work.  I see now that the API in the most recently published 0.4.6 has changed on master.  This looks like it changed a few days ago in https://github.com/fairlearn/fairlearn/pull/424/files#diff-97014611344adf9185e7249ff4c43f4dR41, perhaps in relation to larger discussion in https://github.com/fairlearn/fairlearn-proposals/pull/9/files#diff-44059dad885525fd6f291b8ee1774e15R235.

So I think at this point I'll leave this be :)  Thanks!

**Comment by romanlutz** [comment]
I *think* this is because `DemographicParity` looks at difference in the selection rate between each group and the overall selection rate, so that can be below your threshold, say `0.05`, even if the difference between the min and max are more. Example: 50% of samples in group A, 50% of samples in group B

threshold is 0.05
selection rate of A is 0.75
selection rate of B is 0.81
overall selection rate is 0.78
demographic parity difference is 0.06 (max-min)
difference between each group and the overall selection rate is 0.03, i.e., it satisfied the constraint.

@MiroDudik please correct me if that doesn't make sense.

**Comment by kevinrobinson** [comment]
From @romanlutz over in https://github.com/fairlearn/fairlearn/issues/460:

> `DemographicParity` compares each individual group's selection rate to the overall selection rate, whereas demographic parity difference looks at min vs. max. This isn't document well and actually came up a few times already. Perhaps this isn't intuitive @MiroDudik

**Comment by yuzhenmao** [comment]
Hi, do you know how to show the debug information with debug logging enabled? I tried to add 
`import logging;
logger = logging.getLogger("fairlearn").setLevel(logging.DEBUG)`
but it cannot show the debug info.
Thanks!
